src/nbl/asset/ISPIRVOptimizer.cpp

```diff
@@ -0,0 +1,85 @@
+#include "nbl/asset/ISPIRVOptimizer.h"
+
+#include "spirv-tools/optimizer.hpp" 
+#include "spirv-tools/libspirv.hpp"
+
+#include "nbl/core/core.h"
+#include "nbl/core/IReferenceCounted.h"
+#include "os.h"
+
+using namespace nbl::asset;
+
+static constexpr spv_target_env SPIRV_VERSION = spv_target_env::SPV_ENV_UNIVERSAL_1_5;
+
+nbl::core::smart_refctd_ptr<ICPUBuffer> ISPIRVOptimizer::optimize(const uint32_t* _spirv, uint32_t _dwordCount) const
+{
+    spvtools::Optimizer opt(SPIRV_VERSION);
+
+    // lots of these taken from https://www.lunarg.com/wp-content/uploads/2020/05/SPIR-V-Shader-Legalization-and-Size-Reduction-Using-spirv-opt_v1.2.pdf
+    opt .RegisterPass(spvtools::CreateMergeReturnPass())
+        .RegisterPass(spvtools::CreateInlineExhaustivePass())
+        .RegisterPass(spvtools::CreateEliminateDeadFunctionsPass())
+        .RegisterPass(spvtools::CreateScalarReplacementPass())
+        .RegisterPass(spvtools::CreateLocalSingleBlockLoadStoreElimPass())
+        .RegisterPass(spvtools::CreateLocalSingleStoreElimPass())
+        .RegisterPass(spvtools::CreateSimplificationPass())
+        .RegisterPass(spvtools::CreateVectorDCEPass())
+        .RegisterPass(spvtools::CreateDeadInsertElimPass())
+        .RegisterPass(spvtools::CreateAggressiveDCEPass())
+        .RegisterPass(spvtools::CreateDeadBranchElimPass())
+        .RegisterPass(spvtools::CreateBlockMergePass())
+        .RegisterPass(spvtools::CreateLocalSingleBlockLoadStoreElimPass())
+        .RegisterPass(spvtools::CreateLocalSingleStoreElimPass())
+        .RegisterPass(spvtools::CreateLocalMultiStoreElimPass())
+        .RegisterPass(spvtools::CreateSimplificationPass())
+        .RegisterPass(spvtools::CreateVectorDCEPass())
+        .RegisterPass(spvtools::CreateDeadInsertElimPass())
+        .RegisterPass(spvtools::CreateRedundancyEliminationPass())
+        .RegisterPass(spvtools::CreateAggressiveDCEPass())
+        //.RegisterPass(spvtools::CreateLoopInvariantCodeMotionPass()) //this completely breaks material compiler
```

Comment:
open an issue with SPIR-V Tools (and probably SPIR-V Cross too?), as soon as I merge this.

Provide them with:
- source VK GLSL
- source SPIR-V
- non optimized GL GLSL
- optimization GL GLSL without loop invariant code motion pass
- optimization GL GLSL with loop invariant code motion pass
- renderdoc capture of mitsuba loader example on Nvidia and Intel GPU

Reply:
the earlier you open an issue/bugreport the earlier they'll have a fix for it

Reply:
https://github.com/KhronosGroup/SPIRV-Tools/issues/4093